server/src/main/java/io/crate/expression/scalar/arithmetic/RoundFunction.java

```diff
@@ -58,5 +67,41 @@ public static void register(ScalarFunctionModule module) {
                 }
             );
         }
+        module.register(
+            scalar(
+                NAME,
+                DataTypes.DOUBLE.getTypeSignature(),
+                DataTypes.INTEGER.getTypeSignature(),
+                DataTypes.DOUBLE.getTypeSignature()
+            ),
+            RoundFunction::roundWithPrecision
+        );
+    }
+
+    private static Scalar<Number, Number> roundWithPrecision(Signature signature, BoundSignature boundSignature) {
+        return new Scalar<>() {
+
+            @Override
+            public Signature signature() {
+                return signature;
+            }
+
+            @Override
+            public BoundSignature boundSignature() {
+                return boundSignature;
+            }
+
+            @Override
+            public Number evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<Number>... args) {
+                Number n = args[0].value();
+                Number nd = args[1].value();
+                if (null == n || null == nd) {
+                    return null;
+                }
+                double val = n.doubleValue();
+                int numDecimals = nd.intValue();
+                return BigDecimal.valueOf(val).setScale(numDecimals, RoundingMode.HALF_UP).doubleValue();
```

Comment:
I think if it's not overflowing, `(round(<val> * 10 ^ numDecimals)) / <val> * 10 ^ numDecimals)` would be faster.

Reply:
> I think if it's not overflowing, (round(<val> * 10 ^ numDecimals)) / <val> * 10 ^ numDecimals) would be faster.

There is both the problem of overflowing and precision loss as e.g. pointed out [here](https://www.baeldung.com/java-round-decimal-number#mathround-method). While one could check overflow and precision to fit within the max safe integer bounds of a double, the checks become cumbersome and harder to maintain. As this would be primarily be a perform optimisation and I'd expect this function to be mostly used for presentation purposes after calculation on a final result set, I think it isn't worth it. Therefore I would suggest to stay with the BigDecimal approach.

Reply:
I slightly agree on using `NUMERIC` instead, any objections? @mfussenegger

Reply:
> > I think if it's not overflowing, (round( * 10 ^ numDecimals)) /  * 10 ^ numDecimals) would be faster.
> 
> There is both the problem of overflowing and precision loss as e.g. pointed out [here](https://www.baeldung.com/java-round-decimal-number#mathround-method). While one could check overflow and precision to fit within the max safe integer bounds of a double, the checks become cumbersome and harder to maintain. As this would be primarily be a perform optimisation and I'd expect this function to be mostly used for presentation purposes after calculation on a final result set, I think it isn't worth it. Therefore I would suggest to stay with the BigDecimal approach.

Is detecting a possible overflow really that complex?
I do not completely follow you argumentation about the actual use-cases of this function as we never know for sure how users are using it. 
But I'm also fine on using your approach now, we can follow up here anytime without a user-facing change

Reply:
As mentioned above this method falls apart for inputs like `round(260.775, 2)` with an expected output of `260.78`

```
100 * 260.775 == 26077.499999999996
Math.round(26077.499999999996) == 26077
26077 / 100 == 260.77 but should be 260.78
```

so no. No easy way to detect loss of precision